Guard `fHavePruned` to avoid potential data race

amadeuszpawlik opened the pull request:
As detailed in #21627, there is a potential data race on `fHavePruned` as one thread could be reading it while another one is writing to it.

Guard `fHavePruned`, lock in `IsBlockPruned` (`FlushStateToDisk` is holding `cs_main` while writing to the variable, so this ensures that the data race cannot occur).

amadeuszpawlik committed:
Guard `fHavePruned` to avoid potential data race

A potential data race has been detected on `fHavePruned` where one
thread is writing to it via `FlushStateToDisk` as another thread is
reading it via `IsBlockPruned`.

MarcoFalke commented:
Looks ok, just two notes:

* Would be nice to move blockstorage away from cs_main to it's own lock, so that blockstorage related RPC don't block the validation thread, which might be busy accepting txs to the mempool.
* The fix is a good first step, but likely incomplete. See https://github.com/bitcoin/bitcoin/issues/17161#issuecomment-543151879

amadeuszpawlik committed:
Guard `fHavePruned` with its own lock

Go from using `cs_main` to `cs_HavePruned` as a mutex for accessing
`fHavePruned`. This so as to not block the validation thread with
blockstorage related RPCs.

amadeuszpawlik committed:
Fix double cs_HavePruned lock in init

amadeuszpawlik commented:
Thanks for the feedback @MarcoFalke
I changed `cs_main` to `cs_HavePruned`; good point.
As for your second point, I'll have to investigate more.

amadeuszpawlik commented on src/init.cpp:
```diff
@@ -1476,6 +1477,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
             try {
                 LOCK(cs_main);
+                LOCK(cs_HavePruned);
```
Unsure wheter to lock the whole block or line 1485 specifically
My reasoning here was that verifying blocks ought to have higher priority than answering RPC requests
